### ISSUE TITLE
generate yaml final file from a bunch of separated files

### ISSUE THREAD

**Issue by daniloab** [asyncapi/generator#568]
Hi, I have a package into my application and for each event file, it will be placed beside it the YAML file. And, from a generate function I want to do a bundle of them and generate a final file with each one placed into one place.

### Files structure
```terminal
src
  scripts
    generateAsyncApi.ts
    asyncApiHeaderConfig.js
    asyncApiHeaderConfig.yml
  asyncapi
    AsyncApi.tsx // asyncapi react component that consume 
  modules
      userSignUp
        userSignUp.ts
        userSignUp.yaml
      userSignIn
        userSignIn.ts
        userSignIn.yaml
```

### userSignUp.yaml
```yaml
channels:
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
components:
  messages:
    UserSignedUp:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user
```

### userSignIn.yaml
```yaml
channels:
  user/signedin:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedIn'
components:
  messages:
    UserSignedIn:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user

```

### async api header info config (asyncapiHeader.yaml)
```yaml
asyncapi: '2.0.0'
id: 'id'
info:
  title: User Event API
  version: '1.0.0'
  description: User Event API Specification
```

So, what I'm looking for is a way to do this by a js function to generates the `.json` and `.yaml` final file with these three pieces of yaml. Without duplicating it.

Example final file:
```yaml
asyncapi: '2.0.0'
id: 'id'
info:
  title: User Event API
  version: '1.0.0'
  description: User Event API Specification
channels:
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
  user/signedin:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedIn'
components:
  messages:
    UserSignedUp:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user
    UserSignedIn:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user
```

## What I have tried
I've been tried to use the `swagger-jsdoc` to generate this final file for me. This workflow's very similar to the workflow for our redoc apis.

The final result being generated is an error: `TypeError: Cannot read property 'messages' of undefined`

So, I think that the swagger-jsdoc does not support this schema of channels.

### How can I do this?
But, for now, we want to go forward with AsyncAPI to make better and robust our event drive.

Here an example of the generating function that we are trying to implement to generate this: https://github.com/daniloab/async-api-react-example/blob/master/src/scripts/generateAsyncApi.ts (already configuring the CRA to run the script that calls the function correctly)

So, all of this to asking two things:

1. can we still use swagger-jsdoc to generates this?
2. If don´t, whats is the properly and right way to do this?

**Comment by MikeRalphson** [comment]
Do you have `$ref`s linking your files together? From your examples, it is not clear.

**Comment by MikeRalphson** [comment]
https://www.npmjs.com/package/oas-resolver may do what you need.

**Comment by derberg** [comment]
May I ask why you need to contact it all into one file in the end? I think I'm missing something. 

I just had some fun with your files, this is how they look like, assuming flat structure of files in single directory:

**asyncapi.yml**

```yml
asyncapi: '2.0.0'
info:
  title: User Event API
  version: '1.0.0'
  description: User Event API Specification
channels:
  user/signedin:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedIn'
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
components:
  messages:
    UserSignedIn:
      $ref: 'userSignIn.yml'
    UserSignedUp:
      $ref: 'userSignUp.yml'
```

**userSignIn.yml**

```yml
payload:
  type: object
  properties:
    displayName:
      type: string
      description: Name of the user
    email:
      type: string
      format: email
      description: Email of the user
```

**userSignUp.yml**

```yml
payload:
  type: object
  properties:
    displayName:
      type: string
      description: Name of the user
    email:
      type: string
      format: email
      description: Email of the user
```


Then I run the generator against main file and get docs generates as our JS Parser can handle and resolve references:
```bash
 ag asyncapi.yml @asyncapi/html-template -o output
```


btw, payload for both messages seems to be the same, so you can just keep it once and reference twice

**Comment by daniloab** [comment]
HI, I prepared a better example and cleaner:

https://github.com/daniloab/bull-named-asyncapi-example

the main idea is: get the all yml files, bundle them and generate a the final file with all yml into it without duplicating info

the expect is a generated file like this below:

```yaml
asyncapi: 2.0.0
info:
  title: Entria Event Driven
  version: 1.0.0
  description: Entria event driven docs
channels:
  user/signedin:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedIn'
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
components:
  messages:
    UserSignedIn:
      payload:
        type: object
        properties:
          id:
            type: string
            description: Id of the user
    UserSignedUp:
      payload:
        type: object
        properties:
          id:
            type: string
            description: Id of the user
```

as I said, the `generateAsyncApi.ts` is using the `swagger-jsdoc´ to try do this, but return errors like: 
- `not support channel`
- `not support message`

I have been looking for the @MikeRalphson answer and Im getting some point with that.

From your answer @derberg I need the end file to use to render at someplace to expose for our users.

With this approach it's like I have pieces of YAML and when runs `yarn asyncapi` it will generate the file final for me.

**Comment by derberg** [comment]
I'm not sure I can help you here. I tried running the project, but the structure of the object where you get an error is super bizarre for me:
```json
{
   "asyncapi":"2.0.0",
   "info":{
      "title":"User Event API",
      "version":"1.0.0"
   },
   "description":" User Event API Specification",
   "swagger":"2.0",
   "paths":{
      "channels":{
         "user/signedin":{
            "subscribe":{
               "message":{
                  "$ref":"#/components/messages/UserSignedIn"
               }
            }
         }
      }
   },
   "definitions":{
      
   },
   "responses":{
      
   },
   "parameters":{
      
   },
   "securityDefinitions":{
      
   },
   "tags":[
      
   ]
}
```

like swagger and asyncapi in one, I'm not sure what you're trying to do with this openapi package 🤷🏼 

> From your answer @derberg I need the end file to use to render at someplace to expose for our users.

You mean you want user to see entire file without `$refs`, so a dereferenced file. So just use https://github.com/APIDevTools/json-schema-ref-parser directly, or our JavaScript Parser that uses json-schema-ref-parser (disadvantage is that you will get some additional props added by the parser). Of course I suggest contributing to JS Parser so it has a helper function that after parsing gives you access to entire document without parser-specific props. 

Anyway, please confirm if my assumption is correct, I mean assumption on what you want to achieve

**Comment by daniloab** [comment]
Hi, actually I'm working on this on the swagger-jsdoc side. They already have support for open API v2 and v3. 

We think that can be good to have support for async API.

I made this pr there to have support for `channels` into yml https://github.com/Surnet/swagger-jsdoc/pull/264 and resolve my problem above.

So, I believe that this is not an AsyncAPI issue anymore. 

I finally got the result that I expect and now  I can generate the yml and I'm using the async API React to render them on our frontend.
